vm/actor/src/builtin/miner/mod.rs

```diff
@@ -415,161 +616,278 @@ impl Actor {
-                    format!("wrong proof type {:?}", params.registered_proof),
+        if params.replace_sector_deadline >= WPOST_PERIOD_DEADLINES {
+            return Err(actor_error!(
+                ErrIllegalArgument,
+                "invalid deadline {}",
+                params.replace_sector_deadline
+            ));
+        }
+
+        #[allow(clippy::absurd_extreme_comparisons)]
+        if params.replace_sector_number >= MAX_SECTOR_NUMBER {
+            return Err(actor_error!(
+                ErrIllegalArgument,
+                "invalid sector number {}",
+                params.replace_sector_number
+            ));
+        }
+
+        // gather information from other actors
+
+        let reward_stats = request_current_epoch_block_reward(rt)?;
+        let power_total = request_current_total_power(rt)?;
+        let deal_weight =
+            request_deal_weight(rt, &params.deal_ids, rt.curr_epoch(), params.expiration)?;
+
+        let newly_vested = rt.transaction(|state: &mut State, rt| {
+            let info = get_miner_info(rt, state)?;
+
+            rt.validate_immediate_caller_is(
+                info.control_addresses
+                    .iter()
+                    .chain(&[info.worker, info.owner]),
+            )?;
+
+            let store = rt.store();
+
+            if params.seal_proof != info.seal_proof_type {
+                return Err(actor_error!(
+                    ErrIllegalArgument,
+                    "sector seal proof {:?} must match miner seal proof type {:?}",
+                    params.seal_proof,
+                    info.seal_proof_type
                 ));
-            };
-            let sec = st
-                .get_precommitted_sector(rt.store(), params.sector_number)
+            }
+
+            let max_deal_limit = deal_per_sector_limit(info.sector_size);
+            if params.deal_ids.len() as u64 > max_deal_limit {
+                return Err(actor_error!(
+                    ErrIllegalArgument,
+                    "too many deals for sector {} > {}",
+                    params.deal_ids.len(),
+                    max_deal_limit
+                ));
+            }
+
+            state
+                .allocate_sector_number(store, params.sector_number)
                 .map_err(|e| {
-                    ActorError::new(
-                        ExitCode::ErrIllegalState,
-                        format!(
-                            "failed to check precommitted sector: {}, {}",
-                            params.sector_number, e
-                        ),
+                    e.wrap(format!(
+                        "failed to allocate sector id {}",
+                        params.sector_number
+                    ))
+                })?;
+
+            // The following two checks shouldn't be necessary, but it can't
+            // hurt to double-check (unless it's really just too
+            // expensive?).
+            let _ = state
```

Comment:
```suggestion
            state
```
`let _` just always opens up to future proof issues